compiler/parser/RdrHsSyn.hs

```diff
@@ -2328,6 +2328,38 @@ asPatPrefixCon varName conName =
     (noLoc $ mkRdrUnqual (mkVarOcc varName))
     (noLoc $ XPat (noLoc $ ConPatIn conName $ PrefixCon [])))
 
+-- | The "-Wunused-matches hack." Construct dummy bindings of the form
+-- '_ = this@T{..}' or '_ = self'. We use this hack to suppress
+-- warnings of the form "Defined but not used: ‘this’".
+dummyBind :: Pat GhcPs -> LHsBind GhcPs
+dummyBind p =
+  let wildCard = XPat $ (noLoc (WildPat noExt))
+      grhss = GRHSs noExt
+        [noLoc $ GRHS noExt []
+          (noLoc (HsVar noExt (patToRdrName p)))]
+        (noLoc emptyLocalBinds)
+  in noLoc $ PatBind noExt wildCard grhss ([], [])
+  where
+    patToRdrName :: Pat GhcPs -> Located RdrName
+    patToRdrName (dL -> L _ (VarPat _ n)) = n
+    patToRdrName (dL -> L _ (AsPat _ n _)) = n
+    patToRdrName _ = error "unexpected: patToRdrName"
```

Comment:
It looks like
```
let (a,b) = (0, 1)
```
is a valid template-level let statement atm which is going to cause this to crash.
I wouldn’t bother trying to support this for now and just keep the current state but we shouldn’t crash.

Reply:
I'll put a comment. There is a filter step that will ensure that case will never fire.

Reply:
Oh I see, then I would recommend to do the map and the filter in one step so there is no way to get this wrong instead of adding a comment explaining why the current implementation is safe.

Reply:
Yep, I'll rejig it.

Reply:
✔️